src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -70915,5 +71101,281 @@ static void G<T>(T? x, object y)
+        }
+
+        [WorkItem(30140, "https://github.com/dotnet/roslyn/issues/30140")]
+        [Fact]
+        public void NullCoalescingAssignment_AssignsState()
+        {
+            var source = @"
+class C
+{
+    object? F = null;
+    void M(C? c1, C c2, C c3)
+    {
+        c1 ??= c2;
+        c1.ToString();
+
+        if (c3.F == null) return;
+
+        c1 = null;
+        c1 ??= c3;
+        c1.F.ToString(); // Warn 1
+
+        c1 = null;
+        c1 ??= c3;
+        c1.ToString();
+        c1.F.ToString(); // Warn 2
+
+        if (c1.F == null) return; // Hidden 3
+        c1 ??= c2; // Hidden 4
+        c1.F.ToString(); // Warn 5
+
+        // We could support this in the future if MakeSlot is made smarter to understand
+        // that the slot of a ??= is the slot of the left-hand side. https://github.com/dotnet/roslyn/issues/32501
+        (c1 ??= c3).F.ToString(); // Warn 6, Hidden 7
+    }
+}";
+
+            var comp = CreateCompilation(source, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(
+                // (14,9): warning CS8602: Possible dereference of a null reference.
+                //         c1.F.ToString(); // Warn 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "c1.F").WithLocation(14, 9),
+                // (19,9): warning CS8602: Possible dereference of a null reference.
+                //         c1.F.ToString(); // Warn 2
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "c1.F").WithLocation(19, 9),
+                // (21,13): hidden CS8606: Result of the comparison is possibly always false.
+                //         if (c1.F == null) return; // Hidden 3
+                Diagnostic(ErrorCode.HDN_NullCheckIsProbablyAlwaysFalse, "c1.F == null").WithLocation(21, 13),
+                // (22,9): hidden CS8607: Expression is probably never null.
+                //         c1 ??= c2; // Hidden 4
+                Diagnostic(ErrorCode.HDN_ExpressionIsProbablyNeverNull, "c1").WithLocation(22, 9),
+                // (23,9): warning CS8602: Possible dereference of a null reference.
+                //         c1.F.ToString(); // Warn 5
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "c1.F").WithLocation(23, 9),
+                // (27,9): warning CS8602: Possible dereference of a null reference.
+                //         (c1 ??= c3).F.ToString(); // Warn 6, Hidden 7
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "(c1 ??= c3).F").WithLocation(27, 9),
+                // (27,10): hidden CS8607: Expression is probably never null.
+                //         (c1 ??= c3).F.ToString(); // Warn 6, Hidden 7
+                Diagnostic(ErrorCode.HDN_ExpressionIsProbablyNeverNull, "c1").WithLocation(27, 10));
+        }
+
+        [WorkItem(30140, "https://github.com/dotnet/roslyn/issues/30140")]
+        [Fact]
+        public void NullCoalescingAssignment_RightStateValidInRightOnly()
```

Comment:
>NullCoalescingAssignment_RightStateValidInRightOnly [](start = 20, length = 51)

Are we testing that left side remains nullable on the right?

Reply:
This test could be valuable since we are adjusting slots manually before visiting the right side.

---
In reply to: [248492236](https://github.com/dotnet/roslyn/pull/32225#discussion_r248492236) [](ancestors = 248492236)